src/main/java/com/linecorp/armeria/client/http/HttpRequestSubscriber.java

```diff
@@ -135,8 +139,23 @@ private void writeFirstHeader() {
                          host, firstHeaders.method().name(), firstHeaders.path());
         logBuilder.attr(RequestLog.HTTP_HEADERS).set(firstHeaders);
 
-        encoder.writeHeaders(ctx, id, streamId(), firstHeaders, false).addListener(this);
-        ctx.flush();
+        switch (firstHeaders.method()) {
+            case GET:
+            case HEAD:
+            case PATCH:
+            case DELETE:
+            case TRACE:
+            case CONNECT:
+                // Do not send any content for the methods which is not supposed to have content.
+                // To be precise, the RFC 2616 does not prohibit sending non-empty content for these methods,
+                // but it is discouraged to do so.
+                // See: http://stackoverflow.com/q/978061
```

Comment:
Would this logic be possible at the object encoding level?

1) Write all headers except for content-related ones on header write.
2) On the first publishing of data, send the required headers and then send the data.

Also if the problem is that reactive streams API is not rich enough, I would ditch it. Generic APIs don't have much value if they can't solve the required use cases - for armeria it is implementing a compliant HTTP server I think.